source/core/slang-uint-set.h

```diff
@@ -80,16 +141,105 @@ class UIntSet
         /// Returns true if set1 and set2 have a same value set (ie there is an intersection)
     static bool hasIntersection(const UIntSet& set1, const UIntSet& set2);
 
-private:
-    enum
+    struct Iterator
     {
-        kElementShift = 5,                              ///< How many bits to shift to get Element index from an index
-        kElementSize = sizeof(Element) * 8,             ///< The number of bits in an element
-        kElementMask = kElementSize - 1,                ///< Mask to get shift from an index
+        friend class UIntSet;
+    private:
+        const List<Element>* context;
+        Index block = 0;
+        Element processedElement = 0;
+        uint64_t LSB = 0;
+
+        void clearLSB()
+        {
+            LSB = bitscanForward(processedElement);
+            processedElement &= processedElement - 1;
+        }
+    public:
+        Iterator(const List<Element>* inContext)
+        {
+            context = inContext;
+        }
+
+        Element operator*()
+        {
+            return Element(LSB + (kElementSize * block));
+        }
+
+        template<typename T>
+        T get()
+        {
+            return T(*(*this));
+        }
+
+        Iterator& operator++()
+        {
+            while (processedElement == 0)
+            {
+                block++;
+                if (block >= context->getCount())
+                {
+                    processedElement = 0;
+                    return *this;
+                }
+                processedElement = (*context)[block];
+            }
+            clearLSB();
+            return *this;
+        }
+        Iterator& operator++(int)
+        {
+            return ++(*this);
+        }
+        bool operator==(const Iterator& other) const
+        {
+            return other.block == this->block
+                && other.processedElement == this->processedElement;
+        }
+        bool operator!=(const Iterator& other) const
+        {
+            return !(other == *this);
+        }
     };
+    Iterator begin() const
+    {
+        Iterator tmp(&m_buffer);
+        if (m_buffer.getCount() == 0)
+            return tmp;
+
+        tmp.processedElement = m_buffer[0];
+        if (tmp.processedElement == 0)
+            tmp++;
+
+        tmp.clearLSB();
+
+        return tmp;
```

Comment:
I am not sure what is happening here.
Should it be more like,
```
        tmp.processedElement = 0;
        for (auto buf : m_buffer)
        {
            if (buf == 0)
                continue;
            tmp.processedElement = buf;
            break;
        }
        return tmp;
```
I am not sure why clearLSB() is needed here.

Reply:
we are getting the first valid set bit stored inside `uint64_t LSB;`

What this means is that we o the following:
1. make `Iterator` context
2. if empty, return (equal to `end()`)
3. if `m_buffer[0]` is 0, we have 0 set bits. Iterate our iterator internally (`tmp++`) to find the next element inside `m_buffer` with a set bit.
    * If the entire buffer is 0's, we will get `processedElement = 0` and `block = m_buffer.getCount()` which is equal to `end()`
4. `tmp.clearLSB();` gets the least significant bit location and caches it for processing with `operator*`.

Reply:
I see. Operator++ can loop.
But aren't you calling clearLSB twice? On in operator++ and another one here?

Reply:
You are correct, this is incorrect (capabilities do not get effected since they must have a element in first buffer to be valid).